tenacity/marshaller.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2016 Julien Danjou
+# Copyright 2016 Joshua Harlow
+# Copyright 2013-2014 Ray Holder
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from concurrent import futures
+
+from monotonic import monotonic as now
+import six
+
+from tenacity import _utils
+
+
+NO_RESULT = object()
+
+
+class DoAttempt(object):
+    pass
+
+
+class DoSleep(float):
+    pass
+
+
+class RetryError(Exception):
+    """Encapsulates the last attempt instance right before giving up."""
+
+    def __init__(self, last_attempt):
+        self.last_attempt = last_attempt
+
+    def reraise(self):
+        if self.last_attempt.failed:
+            raise self.last_attempt.result()
+        raise self
+
+    def __str__(self):
+        return "RetryError[{0}]".format(self.last_attempt)
+
+
+class TryAgain(Exception):
+    """Always retry the executed function when raised."""
+
+
+class Future(futures.Future):
+    """Encapsulates a (future or past) attempted call to a target function."""
+
+    def __init__(self, attempt_number):
+        super(Future, self).__init__()
+        self.attempt_number = attempt_number
+
+    @property
+    def failed(self):
+        """Return whether a exception is being held in this future."""
+        return self.exception() is not None
+
+    @classmethod
+    def construct(cls, attempt_number, value, has_exception):
+        """Helper (for testing) for making these objects easily."""
+        fut = cls(attempt_number)
+        if has_exception:
+            fut.set_exception(value)
+        else:
+            fut.set_result(value)
+        return fut
+
+
+class Marshaller(object):
+    def __init__(self, retrying, fn):
+        self.retrying = retrying
+        self.fn = fn
+
+        retrying.statistics.clear()
```

Comment:
Good point. But marshaller is a volatile objects while statistics must remain. I investigate this.